jupyterhub/user.py

```diff
@@ -301,6 +301,84 @@ def sync_groups(self, group_names):
             self.orm_user.groups = []
         self.db.commit()
 
+    def sync_roles(self, auth_roles):
+        """Synchronize roles with database"""
+        auth_roles_by_name = {role['name']: role for role in auth_roles}
+
+        current_user_roles = {r.name for r in self.orm_user.roles}
+        new_user_roles = set(auth_roles_by_name.keys())
+
+        granted_roles = new_user_roles.difference(current_user_roles)
+        stripped_roles = current_user_roles.difference(new_user_roles)
+
+        if granted_roles:
+            self.log.info(f"Granting user {self.name} roles(s): {granted_roles}")
+        if stripped_roles:
+            self.log.info(f"Stripping user {self.name} roles(s): {stripped_roles}")
+
+        all_roles = {r.name for r in self.db.query(orm.Role).all()}
```

Comment:
This might be costly to fetch all roles that exist every time, when most won't be relevant. Instead, you can do a filtered query for:

```python
self.db.query(orm.Role).filter(orm.Role.name.in_(granted_roles))
```

to retrieve only the roles that are to be granted and already exist (or a similar query if you're going to look at different collections)

Reply:
Done in 26906cca07c18e2e20b5b6c6d7ee134876ad2832.